server/src/main/java/io/crate/interval/IntervalParser.java

```diff
@@ -135,6 +139,9 @@ static Period roundToPrecision(Period period, Precision start, Precision end) {
             if (end == Precision.SECOND) {
                 return period.withMillis(0);
             }
+            if (end == Precision.MILLISECOND) {
+                return period.withMillis(0);
+            }
```

Comment:
We should reset next unit to 0, not the `end` one. 
`UNIT1 TO UNIT2` means that we keep precision up to `UNIT2`


Since millis is the smallest unit, we should keep millis -> we should do nothing here and revert changes in the `IntervalParser. roundToPrecision` method. 

Technically speaking, we should replace 
```
if (end == Precision.MILLISECOND) {
    return period.withMillis(0);
}
```

with smth like
```
if (end == Precision.MILLISECOND) {
    return period.withMicros(0);
}
```

But Period doesn't have micros granularity, millis is already the smallest unit so we have nothing to reset.

BTW, not related to this PR but just loudly thinking:

A consequence of Period not having mircoseconds is that we won't be able to do `UNIT TO MILLISECONDS` when when we eventually support `MICROSECOND` unit for INTERVAL - we would have to reset microseconds but won't be able to do so - but now all good as we don't support it.

 `Period` doesn't have `microsecond` unit, so , we would have to reset

Reply:
We need more tests with `SOME_UNIT TO MILLISECONDS` in the `IntervalParserTest` to cover this. 

`IntervalAnalysisTest` tests don't cover it, it doesn't call IntervalParser methods.

Reply:
When checking in the [Postgres manual](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT:~:text=The%20interval%20type%20has%20an%20additional%20option%2C%20which%20is%20to%20restrict%20the%20set%20of%20stored%20fields%20by%20writing%20one%20of%20these%20phrases%3A), it says 

> The interval type has an additional option, which is to restrict the set of stored fields by writing one of these phrases
```
YEAR
MONTH
DAY
HOUR
MINUTE
SECOND
YEAR TO MONTH
DAY TO HOUR
DAY TO MINUTE
DAY TO SECOND
HOUR TO MINUTE
HOUR TO SECOND
MINUTE TO SECOND
```

So probably MILLISECONDS should not be in there at all.

Also, if I test the behavior in Postgres, the second fractions behave differently, like you allueded to above:

```
# select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4.4 seconds' HOUR TO SECOND;
             interval
----------------------------------
 4 years 3 mons 2 days 01:02:04.4

teslamate=# select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4 seconds 400 ms' MINUTE TO SECOND;
             interval
----------------------------------
 4 years 3 mons 2 days 01:02:04.4
(1 row)

teslamate=# select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4 seconds 400 ms' SECOND;
             interval
----------------------------------
 4 years 3 mons 2 days 01:02:04.4
(1 row)
```

Basically seconds are treated as floating point numbers, and ms are not cut off.


Should I change the implementation to reflect this?

Reply:
For completeness here is what crate outputs:
```
cr> select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4.4 seconds' HOUR TO SECOND;
+--------------------------------+
| 'P4Y3M2DT1H2M4S'::interval     |
+--------------------------------+
| 4 years 3 mons 2 days 01:02:04 |
+--------------------------------+
SELECT 1 row in set (0.014 sec)
cr> select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4.4 seconds' DAY TO SECOND;
+--------------------------------+
| 'P4Y3M2DT1H2M4S'::interval     |
+--------------------------------+
| 4 years 3 mons 2 days 01:02:04 |
+--------------------------------+
```

The milliseconds are unset/zero.

Reply:
I thought millis are kept only when they are part of seconds (fractional part of X sec) but even integer sec integer ms keeps millis in Postgres 15:

> select INTERVAL '4 years 3 months 2 days 1 hour 2 minute 4 seconds 2 ms' DAY TO SECOND;
> {"years":4,"months":3,"days":2,"hours":1,"minutes":2,"seconds":4,"milliseconds":2}

So, regarding `TO MILLISECONDS`, I agree that it shouldn't be even supported.

And regarding CrateDB truncating milliseconds in TO SECOND - I would rather create a dedicated issue for that in order to keep the scope of this PR small. We don't always behave 1:1 with PG, so would also need to dicsuss wheter we treat it as a bug, or do it as a feature or don't do at all - all there is not really relevant for MILLISECOND unit task.

Reply:
@BaurzhanSakhariev for the last item I created <a href='https://github.com/crate/crate/issues/16338'>INTERVAL <interval> <UNIT> TO SECOND expressions truncate milliseconds, unlike Postgres · Issue #16338 · crate/crate</a> and removed the code implementing the change to leaving the ms intact.

Reply:
Thanks!